### PR TITLE
SLING-10030 - Addin PostOperation ranking

### PR DESCRIPTION
**Pull request by tomasalmeida**
- Migrating SlingPostServletTest to Junit4
- Creating an auxiliary SortedByRankingList class which stores elements by a given ranking
- Storing all bound PostOperations by ranking and using the one with higher ranking
   - if no ranking is provided, order follows the behavior already established behavior: last Operation bound is used as default one.

Full proposal is described in https://issues.apache.org/jira/browse/SLING-10030

### CODE REVIEW

File: src/main/java/org/apache/sling/servlets/post/impl/helper/SortedByRankingList.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
+/*
```
Comment (paul-bjorkstrand) [inline src/main/java/org/apache/sling/servlets/post/impl/helper/SortedByRankingList.java:1]: This looks like it could be replaced with an implementation of [SortedMap] rather than a new, custom class. A benefit of using a SortedMap is that you can use either a concurrency safe version or not, depending on the use case.

[SortedMap]: https://docs.oracle.com/javase/8/docs/api/java/util/SortedMap.html

Comment (tomasalmeida) [reply src/main/java/org/apache/sling/servlets/post/impl/helper/SortedByRankingList.java:1]: Hi Paul, great point, I kept the same insertion code used before just moving it to the new Class, so it could reused.

I can reimplement it using a SortedMap as proposed. Although I am not sure we can guarantee if no ranking is provided, the last inserted element will be the chosen one.

Let me know if I should provide the implementation using a SortedMap and I can provide it.

Comment (rombert) [reply src/main/java/org/apache/sling/servlets/post/impl/helper/SortedByRankingList.java:1]: I would check what the OSGi core bundles do - I think that the 'sort by ranking' problem occurs there as well. @cziegeler / @karlpauls / @bosschaert  - any suggestions?

Comment (kwin) [reply src/main/java/org/apache/sling/servlets/post/impl/helper/SortedByRankingList.java:1]: Please rather use DS field injection which orders automatically by ranking

> For a reference with multiple cardinality, a field must be a collection of one of the following types:
Collection
List
A subtype of Collection - This type can only be used for dynamic references using the update reference field option. The component instance must initialize the field to a collection object in its constructor.
(https://docs.osgi.org/specification/osgi.cmpn/7.0.0/service.component.html#service.component-field.injection)

and
> For a reference with multiple cardinality, the type of the field must be Collection or List. If the field has a different type, SCR must log an error message with the Log Service, if present, and the field must not be modified. Before the component instance is activated, SCR must set the field value with a new mutable collection that must contain the initial set of bound services sorted using the same ordering as ServiceReference.compareTo based upon service ranking and service id. The collection may be empty if the reference has optional cardinality and there are no bound services. If the reference is dynamic, when there is a change in the set of bound services or the service properties of a bound service are modified and the collection holds service properties, SCR must replace the field value with a new mutable collection that must contain the updated set of bound services sorted using the same ordering as ServiceReference.compareTo based upon service ranking and service id. The new collection may be empty if the reference has optional cardinality and there are no bound services.

(https://docs.osgi.org/specification/osgi.cmpn/7.0.0/service.component.html#d0e37828

Comment (tomasalmeida) [reply src/main/java/org/apache/sling/servlets/post/impl/helper/SortedByRankingList.java:1]: Thank you for the feedback! I will rework on this PR and reimplement the binding using the references given. I believe it is straightforward to implement it for NodeNameGenerator, SlingPostProcessor and PostResponseCreator lists.

Although, I need to rethink how to bind/unbind the postOperation based on OperationName, as in the implementation we have a Map(operationName, List of PostOperation services). Having one large list is not optimal and the map was useful.

If any of you have an idea how I can solve the map problem, I accept suggestions 😄

File: src/test/java/org/apache/sling/servlets/post/impl/SlingPostServletTest.java [hunk 2]
```diff
@@ -238,31 +355,31 @@ void setStatusParam(String statusParam) {
...
-		
-		void setSendError(String sendErrorParam) {
-		    this.sendError = sendErrorParam;
-		}
-		
-		public <AdapterType> AdapterType adaptTo(Class<AdapterType> type) {
-		    return null;
-		}
-	}
+            MockSlingHttpServlet3Request {
```
Comment (rombert) [inline src/test/java/org/apache/sling/servlets/post/impl/SlingPostServletTest.java:358]: Please avoid whitespace-only changes, or separate them to their own commits.
